common/physics-system.cpp

```diff
@@ -93,21 +102,49 @@ namespace tec {
 				body->clearForces();
 			}
 
+			// handle changes to desired deactivation mode
 			if (collidable->disable_deactivation) {
 				body->forceActivationState(DISABLE_DEACTIVATION);
 			}
 			else if (body->getActivationState() == DISABLE_DEACTIVATION) {
 				body->forceActivationState(ACTIVE_TAG);
 			}
 
+			// prevent the simulation from rotating the object
+			// this doesn't account for change after creation, once disabled there isn't a re-enable
 			if (collidable->disable_rotation) {
 				body->setAngularFactor(btVector3(0.0, 0, 0.0));
 			}
-			body->setWorldTransform(collidable->motion_state.transform);
-			this->dynamicsWorld->addRigidBody(body);
 
-			if (state.velocities.find(entity_id) != state.velocities.end()) {
-				const Velocity& vel = state.velocities.at(entity_id);
+			// here we add the body to the world if it's not yet
+			// this can later expand to handling multiple dynamics worlds if need be
+			if (!collidable->in_world) {
+				// snap the body to it's position when we add it
+				body->setWorldTransform(collidable->motion_state.transform);
+				collidable->in_world = true;
+				this->dynamicsWorld->addRigidBody(body);
+			}
+			else {
+				btTransform& body_transform = body->getWorldTransform();
+				// simulation motion estimation lite
+				// on the server, this doesn't really do anything
+				// on the client however, this smooths out the motion between the local estimation and server state
+				// this provides a crude but effective error correction until a better one is added ;)
+				if (body_transform.getOrigin().distance(collidable->motion_state.transform.getOrigin()) > 0.01) {
+					// FIXME there is a race condition outside of PhysicsSystem, where the position of the entity
+					// is momentarily at origin (0,0,0) during entity creation (affects server)
+					body->translate(0.5 * (collidable->motion_state.transform.getOrigin() - body_transform.getOrigin()));
```

Comment:
the 0.5 is the error correction curve factor, I'm indecisive about it, but 0.5 seems like a decent compromise, smaller numbers are smoother, but also make controls feel more "floaty". Simply snapping to the error target doesn't always work great either.